src/Chapter3.hs

```diff
@@ -1006,7 +1155,126 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
-
+class FightAction action fighter => Fighter fighter action | fighter -> action where
+    health :: fighter -> Health
+    attack :: fighter -> Attack
+    actions :: fighter -> [action]
+    setHealth :: fighter -> Health -> fighter
+    isDead :: fighter -> Bool
+    takeDamage :: fighter -> Attack -> fighter
+    takeDamage fighter attack = setHealth fighter (calculateAttack (health fighter) attack)
+
+class Fighter fighter action => FightAction action fighter where
+    perform :: (Fighter f a) => action -> fighter -> f -> (fighter, f)
```

Comment:
WOW 😮 mutually recursive typeclasses, never seen such architecture before. Very interesting approach. Very huge work with this, bravo 👏🏼 

I think it is a bit overcomplicated and can be avoided generally. 
I would personally not include `actions` function into the typeclass and avoing using such complicated features. It is actually the only thing you use all the `UndecidableSupeeClasses` for, so I thing the cost is too high.

As you mentioned, you could provide two lists of actions for each fighter separately into the fight function, or you can still have it in the data type, but just not include that function into the typeclass 🙂 Either way is okay and would simplify the solution. Especially as I see, you do exactly that (put lists of actions separately) in the `battle` function, so the only thing the advanced features are used in is in the act of taking the list of actions, which is not that essential to pay that much for 🙂

Reply:
Yeah, didn't like this idea from the beginning 😅
I simplified my solution a bit. I'm still not happy with it, but really want to keep Alive and Dead types 🤷‍♀️